### ISSUE TITLE
Javascript insertAt API lets you get document into a confusing state

### ISSUE THREAD

**Issue by ConradIrwin** [automerge/automerge#602]
I'm building a VSCode extension that keeps an automerge document up to date, I ended up doing this:

```
    this.doc = automerge.change(this.doc, d => {
      for (let c of changes) {
          d.content.deleteAt(c.rangeOffset, c.rangeLength)
          d.content.insertAt(c.rangeOffset, c.text))
      }
    })
```

This works fine for single-character changes, but there's a bug. For `insertAt` I need to actually do something like: `d.content.insertAt(c.rangeOffset, ...c.text.split('')))` because I need to pass each character one at a time to automerge otherwise the offsets into the document get messed up.

A more reduced example is this:

```ts
import * as automerge from '@automerge/automerge'
let doc = automerge.init<{content: automerge.Text}>()
doc = automerge.change(doc, d => {
  d.content = new automerge.Text()
  d.content.insertAt(0, 'ten')
})

console.log(doc.content.toString()) // => "ten"
console.log(doc.content.length) // => 1 !!!!
```

I would expect the previous code to act as though I had done this instead:
```
import * as automerge from '@automerge/automerge'
let doc = automerge.init<{content: automerge.Text}>()
doc = automerge.change(doc, d => {
  d.content = new automerge.Text()
  d.content.insertAt(0, 't', 'e', 'n')
})

console.log(doc.content.toString()) // => "ten"
console.log(doc.content.length) // => 3
```

As I understand it the goal is to have one index per unicode codepoint (I wrote up my understanding [here](https://pkg.go.dev/github.com/automerge/automerge-go#Text)). It turns out this isn't very intuitive in javascript!

What I'd like to make work is:
```
d.content.insertAt(0, '🧟‍♀️')
```

This should insert four unicode codepoints (U+1FD9F (Zombie), U+200D (zero-width joiner), U+2640 (female), U+FE0F (variation selector)):

This took me a while to find
* `'🧟‍♀️'.split()` gives a list of length one: '🧟‍♀️'; 
* `'🧟‍♀️'.split('')` gives a list of length five:  (U+D83E (first surrogate), U+DDDF (second surrogate), U+200D (zero-width joiner), U+2640 (female), U+FE0F (variation selector)).
* [...'🧟‍♀️'] gives the list I need (though I would not have thought to try that without [this](https://stackoverflow.com/questions/54369513/how-to-count-the-correct-length-of-a-string-with-emojis-in-javascript)).

In summary I think we should change the API of the text .insertAt() so that `d.content.insertAt(0, '🧟‍♀️')` acts the same as `d.content.insertAt(0, ...'🧟‍♀️')`, and correctly ensures that the document has one index per unicode codepoint.

At the same time I think we should change the API to throw an error if you do `d.content.insertAt(0, ...'🧟‍♀️'.split(''))` because this ends up inserting two surrogate pairs into the document, which come back out as unicode replacement characters instead of a zombie.

Even if we do fix this, there's still an issue for users to figure out the correct index to use, but that seems solvable separately.

**Comment by ept** [comment]
I'm supportive of this API change. Allowing users to insert multiple code points as a single list element is an unnecessary footgun. Also agree that inserting half of a surrogate pair should throw an error.

For figuring out the correct index to use, I believe Automerge already has the ability to calculate indexes based on UTF-16 code units, which would make it compatible with the indexing in JavaScript strings. Not sure what the exact status of this is, though.

Incidentally, is your VSCode extension generally available? I've been using @pvh's [vscode-automerge-tracker](https://marketplace.visualstudio.com/items?itemName=inkandswitch.vscode-automerge-tracker) while writing a paper to record an editing trace as an Automerge document, but that extension requires very careful use to avoid the VSCode state and the Automerge state becoming inconsistent with each other. If you have something more robust, I'd love to use it!

**Comment by ConradIrwin** [comment]
@ept I don't (yet) have anything more robust, as my extension does roughly the same thing as https://github.com/pvh/vscode-automerge-tracker/blob/main/src/extension.ts#L40-L41.

I know I have to solve the utf-16 code units -> unicode code points indexing problem, are there other edge cases that you have to avoid when working with pvh's extension that I could pre-emptively fix?

**Comment by ept** [comment]
It's been a few months since I last used the VSCode extension so I can't quite remember. I think one issue was that if the file was modified outside of VSCode (e.g. through `git pull` bringing in changes from a collaborator), the tracker didn't notice. Another issue (IIRC) was that if you restarted VSCode, tabs that were automatically reopened after the restart didn't reliably pick up input events; sometimes I had to manually close and re-open a tab for it to start tracking input. But I might be misremembering. I resorted to having a little CLI tool that compared the state of the Automerge doc to the state of the plain text file, and I ran that tool frequently to make sure the two were still in sync.

**Comment by ConradIrwin** [comment]
Makes sense, thanks. (I also have seen the VScode restart issue; but have not yet implemented syncing so haven't got those yet :D)

**Comment by ConradIrwin** [comment]
After talking this through at the meeting yesterday; I'm going to add a patch to this method to fix the footgun herein; but not spend much time on any more concrete APIs (e.g. to fix the utf-16 <> codepoint mapping problem).

The current "unstable" API exposes a new mechanism for text in Automerge.splice that handles indexes javascript-natively; and in the meantime you can use https://npmjs.com/package/code-point-mapping to fix this in the meantime.
